### PR TITLE
Use range-based for loops in a variety of places.

### PR DESCRIPTION
**Pull request by bangerth**
Addresses a number of those places that could benefit from #3774 . Required some wicked perl magic + more manual labor than I wanted this to be :-)

I've put thought into whether to capture the variables by-value or by-reference. I've generally chosen references if it's not a scalar, but by value if it's an integer of some type.

### CODE REVIEW

**Comment by bangerth** [comment]
Oh, it breaks 260 tests... This will have to wait for a few days, I'm afraid :-(

**Comment by tjhei** [comment]
> Oh, it breaks 260 tests...

candidate for the bucket award? :-)

**Comment by tjhei** [comment]
Fascinating. How on earth can range-based for loops break the tests?

**Comment by tjhei** [comment]
This looks like you introduced a bug:
```

##21      <==      Heat fluxes through boundary parts: 1.176e+06 W, -4.157e+06 W, 0 W, 0 W
##21      ==>      Heat fluxes through boundary parts: 1.176e+06 W, 1.176e+06 W, 1.176e+06 W, 1.176e+06 W, 

```

File: source/postprocess/heat_flux_statistics.cc [hunk 1]
```diff
@@ -61,37 +61,31 @@ namespace aspect
...
         Utilities::MPI::sum (local_values, this->get_mpi_communicator(), global_values);
 
         // and now take them apart into the global map again
         unsigned int index = 0;
-        for (std::set<types::boundary_id>::const_iterator
-             p = boundary_indicators.begin();
-             p != boundary_indicators.end(); ++p, ++index)
-          global_boundary_fluxes[*p] = global_values[index];
+        for (const auto p : boundary_indicators)
+          global_boundary_fluxes[p] = global_values[index];
```
Comment (tjhei) [inline source/postprocess/heat_flux_statistics.cc:74]: and this seems to be a bug.

File: source/postprocess/heat_flux_statistics.cc [hunk 2]
```diff
@@ -61,37 +61,31 @@ namespace aspect
...
       }
 
       // now add all of the computed heat fluxes to the statistics object
       // and create a single string that can be output to the screen
       std::ostringstream screen_text;
       unsigned int index = 0;
-      for (std::map<types::boundary_id, double>::const_iterator
-           p = global_boundary_fluxes.begin();
-           p != global_boundary_fluxes.end(); ++p, ++index)
```
Comment (tjhei) [inline source/postprocess/heat_flux_statistics.cc:87]: same here

**Comment by bangerth** [comment]
On 8/13/20 8:39 AM, Timo Heister wrote:
> candidate for the bucket award? :-)

Yep, I'm afraid so :-(

File: source/mesh_refinement/interface.cc [hunk 3]
```diff
@@ -159,13 +157,11 @@ namespace aspect
       std::vector<Vector<float> > all_error_indicators (mesh_refinement_objects.size(),
                                                         Vector<float>(error_indicators.size()));
       unsigned int index = 0;
-      for (typename std::list<std::unique_ptr<Interface<dim> > >::const_iterator
-           p = mesh_refinement_objects.begin();
-           p != mesh_refinement_objects.end(); ++p, ++index)
```
Comment (tjhei) [inline source/mesh_refinement/interface.cc:164]: this is incorrect

File: source/particle/property/interface.cc [hunk 4]
```diff
@@ -365,10 +362,9 @@ namespace aspect
         particle_properties.reserve(property_information.n_components());
 
         unsigned int property_index = 0;
-        for (typename std::list<std::unique_ptr<Interface<dim> > >::const_iterator
-             p = property_list.begin(); p!=property_list.end(); ++p, ++property_index)
```
Comment (tjhei) [inline source/particle/property/interface.cc:369]: this is incorrect

File: source/postprocess/heat_flux_densities.cc [hunk 5]
```diff
@@ -93,16 +93,14 @@ namespace aspect
       // and create a single string that can be output to the screen
       std::ostringstream screen_text;
       unsigned int index = 0;
-      for (std::map<types::boundary_id, double>::const_iterator
-           p = global_boundary_flux_densities.begin();
-           p != global_boundary_flux_densities.end(); ++p, ++index)
```
Comment (tjhei) [inline source/postprocess/heat_flux_densities.cc:98]: this is incorrect

File: source/postprocess/heating_statistics.cc [hunk 6]
```diff
@@ -100,11 +100,9 @@ namespace aspect
               local_mass += out.densities[q] * fe_values.JxW(q);
 
             unsigned int index = 0;
-            for (typename std::list<std::unique_ptr<HeatingModel::Interface<dim> > >::const_iterator
-                 heating_model = heating_model_objects.begin();
-                 heating_model != heating_model_objects.end(); ++heating_model, ++index)
```
Comment (tjhei) [inline source/postprocess/heating_statistics.cc:105]: this is incorrect

File: source/postprocess/mass_flux_statistics.cc [hunk 7]
```diff
@@ -104,37 +104,31 @@ namespace aspect
...
         // then collect contributions from all processors
         std::vector<double> global_values (local_values.size());
         Utilities::MPI::sum (local_values, this->get_mpi_communicator(), global_values);
 
         // and now take them apart into the global map again
         unsigned int index = 0;
-        for (std::set<types::boundary_id>::const_iterator
-             p = boundary_indicators.begin();
-             p != boundary_indicators.end(); ++p, ++index)
-          global_boundary_fluxes[*p] = global_values[index];
```
Comment (tjhei) [inline source/postprocess/mass_flux_statistics.cc:121]: this is incorrect

File: source/postprocess/mass_flux_statistics.cc [hunk 8]
```diff
@@ -104,37 +104,31 @@ namespace aspect
...
       }
 
       // now add all of the computed mass fluxes to the statistics object
       // and create a single string that can be output to the screen
       std::ostringstream screen_text;
       unsigned int index = 0;
-      for (std::map<types::boundary_id, double>::const_iterator
-           p = global_boundary_fluxes.begin();
-           p != global_boundary_fluxes.end(); ++p, ++index)
+      for (const auto &p : global_boundary_fluxes)
```
Comment (tjhei) [inline source/postprocess/mass_flux_statistics.cc:124]: this is incorrect

File: source/postprocess/velocity_boundary_statistics.cc [hunk 9]
```diff
@@ -143,13 +139,11 @@ namespace aspect
         // and now take them apart into the global map again
         // At the same time, calculate the rms velocity for each boundary
         unsigned int index = 0;
-        for (std::set<types::boundary_id>::const_iterator
-             p = boundary_indicators.begin();
-             p != boundary_indicators.end(); ++p, ++index)
+        for (const auto p : boundary_indicators)
```
Comment (tjhei) [inline source/postprocess/velocity_boundary_statistics.cc:142]: this is incorrect

File: source/postprocess/visualization/heating.cc [hunk 10]
```diff
@@ -144,11 +144,9 @@ namespace aspect
           AssertThrow(false, ExcNotImplemented());
 
         unsigned int index = 0;
-        for (typename std::list<std::unique_ptr<HeatingModel::Interface<dim> > >::const_iterator
-             heating_model = heating_model_objects.begin();
-             heating_model != heating_model_objects.end(); ++heating_model, ++index)
+        for (const auto &heating_model : heating_model_objects)
```
Comment (tjhei) [inline source/postprocess/visualization/heating.cc:147]: here you can delete the index.

**Comment by bangerth** [comment]
Thanks for your help with finding all of these places. Let's see how this is going to shake out now. If I had been a bit smarter, I wouldn't have deleted everything after the `for (...variable = abc.begin(); ...` and instead checked whether there is also an `index` variable somewhere...

**Comment by bangerth** [comment]
Only 31 tests failing now :-)

**Comment by bangerth** [comment]
Down to 19 failures, out of 811. Surely that's good enough to merge, right? I mean, that's a 97.66% success rate!

File: source/termination_criteria/interface.cc [hunk 11]
```diff
@@ -90,13 +88,11 @@ namespace aspect
       // call the execute() functions of all plugins we have
       // here in turns.
       std::list<std::string>::const_iterator  itn = termination_obj_names.begin();
-      for (typename std::list<std::unique_ptr<Interface<dim> > >::const_iterator
-           p = termination_objects.begin();
-           p != termination_objects.end(); ++p,++itn)
```
Comment (tjhei) [inline source/termination_criteria/interface.cc:95]: here is another bug.

Comment (bangerth) [reply source/termination_criteria/interface.cc]: Thanks. You don't have to find these bugs for me, though I very much appreciate that you do :-)

**Comment by bangerth** [comment]
Oh, now, look at this: It actually works! Although I will 'fess up to the fact that that took far more work than I actually wanted to put into this patch...

**Comment by tjhei** [comment]
So, did you just exclude the places where you made mistakes before?

**Comment by bangerth** [comment]
On 8/15/20 1:36 PM, Timo Heister wrote:
> So, did you just exclude the places where you made mistakes before?

Yes. There is really not very much one can do about the issue of iterating 
over a collection while also incrementing an index: This can't easily be fit 
into a range-based for loop unless it's a random-access collection for which 
one could use |operator-|. (There is boost's zip-iterator, but that seemed 
like a step backward in readability.)

**Pull request merged**